File: sdks/node/src/managed-sink.js
// Bounded managed candidate sink with fail-open delivery.
//
// Mirrors crates/reproit-sdk-rust/src/managed_sink.rs: a bounded in-process
// queue, one sequential delivery worker, recall counters without customer
// values, and fail-open semantics. A managed SDK failure never changes the
// application's behavior.

import { Buffer } from "node:buffer";

import {
  MAX_GLOBAL_BYTES,
  MAX_OPERATION_BYTES,
  MAX_QUEUED_CANDIDATES,
  canonicalBytes,
} from "./index.js";
import {
  FrozenManagedCaptureClosure,
  PreparedManagedCandidate,
} from "./managed-candidate.js";
import {
  ManagedWorkloadIdentityState,
} from "./managed-identity.js";
import {
  ManagedError,
  canonicalDigest,
  decodeBase64url,
  encodeBase64url,
  isObject,
  nowTimestamp,
  requireTypedId,
  schemaInvalid,
  signBytes,
  validTimestamp,
  validTypedId,
  validateCapabilities,
  verificationKey,
  verifySignedValue,
} from "./managed-protocol.js";
import {
  managedWorkloadKeyId,
} from "./managed-transport.js";
import {
  packageRunningNodeSubject,
  subjectBinding,
} from "./managed-subject.js";
import { captureProcessorCapabilities } from "./processor-capture.js";

export const REGISTRATION_TIMEOUT_MS = 5_000;
export const CANDIDATE_DELIVERY_LIFETIME_MS = 1_000;
const COUNTER_MAXIMUM = Number.MAX_SAFE_INTEGER;
const RECALL_KEYS = [
  "candidate_delivery_expired",
  "candidate_durably_accepted",
  "candidate_incomplete",
  "candidate_queue_full",
  "candidate_rejected",
  "eligible_failure_observed",
  "suppressed_exact_storm",
  "suppressed_high_cardinality_storm",
];

// Deliver complete managed candidates through the bounded upload session.
// Construct with the async factory: ManagedCandidateSink.create(...).
export class ManagedCandidateSink {
  #active = false;
  #client;
  #closure;
  #configuration;
  #operationId;
  #processing = false;
  #queue = [];
  #queuedBytes = 0;
  #queuedCandidates = 0;
  #recall = Object.fromEntries(RECALL_KEYS.map((key) => [key, 0]));
  #registration;
  #subject;
  #signedDeployment;
  #workloadSigningKey;
  #workloadKeyId;
  #workloadPublicKey;
  #worldId;
  // The delivery lifetime is a protocol constant. Tests may shorten it.
  _deliveryLifetimeMs = CANDIDATE_DELIVERY_LIFETIME_MS;

  constructor(
    token,
    client,
    closure,
    configuration,
    subject,
    signedDeployment,
    workloadSigningKey,
    registration,
    options,
  ) {
    if (token !== constructionToken) {
      throw new TypeError(
        "Use ManagedCandidateSink.create to construct the managed sink.",
      );
    }
    this.#client = client;
    this.#closure = closure;
    this.#configuration = {
      captureSignerId: configuration.captureSignerId,
      captureSignerPublicKey: configuration.captureSignerPublicKey,
      serviceId: configuration.serviceId,
    };
    this.#registration = registration;
    this.#subject = subject;
    this.#workloadSigningKey = workloadSigningKey;
    this.#operationId = options.operationId ?? null;
    this.#worldId = closure.worldId();
    this.#workloadPublicKey = verificationKey(workloadSigningKey);
    this.#workloadKeyId = options.workloadKeyId;
    this.#signedDeployment = structuredClone(signedDeployment);
  }

  // Validate the configuration, freeze the closure, package the running
  // subject when none is supplied, and register the signed deployment.
  static async create(client, closure, configuration, options = {}) {
    validateConfiguration(configuration);
    if (!(closure instanceof FrozenManagedCaptureClosure)) {
      closure = new FrozenManagedCaptureClosure(closure);
    }
    if (options.operationId !== undefined && options.operationId !== null) {
      requireTypedId(options.operationId, "operation_id");
    }
    const subject = options.subject ?? packageRunningNodeSubject();
    if (!isObject(options.deployment)) {
      throw schemaInvalid("The managed deployment is required.");
    }
    const deployment = options.deployment;
    prepareManagedDeployment(deployment, subject.manifest, configuration.serviceId);
    validateUnsignedDeployment(deployment);
    const bindingDigest = managedDeploymentBindingDigest(deployment);
    const workloadState = options.workloadStateRoot === undefined
      ? ManagedWorkloadIdentityState.fromEnvironment(bindingDigest)
      : new ManagedWorkloadIdentityState(options.workloadStateRoot, bindingDigest);
    const workloadSigningKey = workloadState.loadOrCreateKey();
    deployment.signed_at = workloadState.loadOrCreateDeploymentSignedAt(
      bindingDigest,
      nowTimestamp(),
    );
    const workloadPublicKey = verificationKey(workloadSigningKey);
    const publicKey = encodeBase64url(workloadPublicKey);
    const workloadKeyId = managedWorkloadKeyId(publicKey);
    deployment.signer_key_id = workloadKeyId;
    deployment.signature = "";
    deployment.signature = signBytes(
      canonicalBytes(deployment),
      workloadSigningKey,
    );
    validateDeployment(deployment);
    const registrationRequest = {
      algorithm: "Ed25519",
      deployment: structuredClone(deployment),
      public_key: publicKey,
      service_id: configuration.serviceId,
    };
    const receipt = {
      deployment_digest: canonicalDigest(deployment),
      service_id: configuration.serviceId,
      workload_key_id: workloadKeyId,
    };
    const registered = workloadState.loadRegistrationReceipt(receipt) !== null;
    return new ManagedCandidateSink(
      constructionToken,
      client,
      closure,
      configuration,
      subject,
      deployment,
      workloadSigningKey,
      {
        projectTokenProvider: projectTokenProvider(configuration),
        receipt,
        registered,
        request: registrationRequest,
        workloadState,
      },
      {
        operationId: options.operationId ?? null,
        workloadKeyId,
      },
    );
  }

  get processingModes() {
    return new Set(["managed"]);
  }

  get queuedBytes() {
    return this.#queuedBytes;
  }

  // Bounded counters that contain no customer values.
  get recallCounters() {
    return { ...this.#recall };
  }

  get subjectManifest() {
    return this.#subject.manifest;
  }

  get workloadKeyId() {
    return this.#workloadKeyId;
  }

  get workloadPublicKey() {
    return this.#workloadPublicKey;
  }

  get worldId() {
    return this.#worldId;
  }

  // Bind the deployment to this subject and sign it as this workload.
  bindDeployment(deployment) {
    prepareManagedDeployment(
      deployment,
      this.#subject.manifest,
      this.#configuration.serviceId,
    );
    if (
      managedDeploymentBindingDigest(deployment) !==
      managedDeploymentBindingDigest(this.#signedDeployment)
    ) {
      throw new ManagedError(
        "AUTHORIZATION_DENIED",
        "The managed deployment does not match the registered deployment.",
      );
    }
    for (const key of Object.keys(deployment)) delete deployment[key];
    Object.assign(deployment, structuredClone(this.#signedDeployment));
  }

  async waitUntilIdle(timeoutMs) {
    const deadline = performance.now() + timeoutMs;
    while (this.#active || this.#queuedCandidates !== 0) {
      const remaining = deadline - performance.now();
      if (remaining <= 0) {
        return false;
      }
      await new Promise((resolve) =>
        setTimeout(resolve, Math.min(10, Math.max(1, remaining))),
      );
    }
    return true;
  }

  // Queue one complete candidate. Never throw into the application.
  trySend(captureId, candidate) {
    let value;
    try {
      value = this.#authorizedCandidate(captureId, candidate);
    } catch {
      this.#increment("candidate_incomplete");
      return false;
    }
    if (
      this.#queuedCandidates >= MAX_QUEUED_CANDIDATES ||
      this.#queuedBytes + candidate.length > MAX_GLOBAL_BYTES
    ) {
      this.#increment("candidate_queue_full");
      return false;
    }
    this.#queue.push({
      enqueued: performance.now(),
      size: candidate.length,
      value,
    });
    this.#queuedBytes += candidate.length;
    this.#queuedCandidates += 1;
    queueMicrotask(() => this.#work());
    return true;
  }

  #authorizedCandidate(captureId, candidate) {
    const bytes = Buffer.from(candidate);
    if (bytes.length > MAX_OPERATION_BYTES) {
      throw schemaInvalid();
    }
    const value = JSON.parse(bytes.toString("utf8"));
    if (
      !isObject(value) ||
      !Buffer.from(canonicalBytes(value)).equals(bytes) ||
      value.capture_id !== captureId ||
      value.processing_mode !== "managed"
    ) {
      throw schemaInvalid();
    }
    const deployment = value.deployment;
    if (
      !isObject(deployment) ||
      deployment.processing_mode !== "managed" ||
      deployment.service_id !== this.#configuration.serviceId ||
      deployment.signer_key_id !== this.#workloadKeyId ||
      (this.#operationId !== null &&
        value.operation_id !== this.#operationId)
    ) {
      throw new ManagedError(
        "AUTHORIZATION_DENIED",
        "The managed deployment does not use the registered workload key.",
      );
    }
    verifySignedValue(deployment, this.#workloadPublicKey);
    return value;
  }

  async #work() {
    if (this.#processing) return;
    this.#processing = true;
    try {
      while (this.#queue.length > 0) {
        const candidate = this.#queue.shift();
        try {
          if (
            performance.now() - candidate.enqueued >=
            this._deliveryLifetimeMs
          ) {
            this.#increment("candidate_delivery_expired");
            continue;
          }
          this.#active = true;
          try {
            await this.#deliver(candidate.value);
            this.#increment("candidate_durably_accepted");
          } catch (error) {
            if (error instanceof ManagedError) {
              this.#recordFailure(error);
            } else {
              this.#increment("candidate_rejected");
            }
          }
        } finally {
          this.#active = false;
          this.#queuedBytes = Math.max(0, this.#queuedBytes - candidate.size);
          this.#queuedCandidates = Math.max(0, this.#queuedCandidates - 1);
        }
      }
    } finally {
      this.#processing = false;
    }
  }

  async #deliver(candidate) {
    const configuration = this.#configuration;
    const prepared = PreparedManagedCandidate.prepareComplete(
      candidate,
      this.#subject,
      this.#closure,
    );
    await this.#ensureRegistered();
    const grant = await prepared.requestEncryptionGrant(
      this.#client,
      this.#workloadKeyId,
      this.#workloadSigningKey,
    );
    const sealed = prepared.seal(
      grant,
      nowTimestamp(),
      configuration.captureSignerId,
      configuration.captureSignerPublicKey,
    );
    try {
      const renewal = await sealed.requestCaptureGrantRenewal(
        this.#client,
        this.#workloadKeyId,
        this.#workloadSigningKey,
      );
      sealed.applyRenewedCaptureGrant(
        renewal,
        nowTimestamp(),
        configuration.captureSignerId,
        configuration.captureSignerPublicKey,
      );
      await sealed.upload(this.#client);
    } finally {
      sealed.dispose();
    }
  }

  async #ensureRegistered() {
    if (this.#registration.registered) return;
    if (this.#registration.projectTokenProvider === null) {
      throw registrationTokenRequired();
    }
    const projectToken = await this.#registration.projectTokenProvider();
    const registration = await this.#client.registerWorkloadKey(
      projectToken,
      this.#registration.request,
      REGISTRATION_TIMEOUT_MS,
    );
    const receipt = this.#registration.receipt;
    if (
      registration.key_id !== receipt.workload_key_id ||
      registration.deployment_digest !== receipt.deployment_digest ||
      registration.service_id !== receipt.service_id
    ) {
      throw new ManagedError(
        "ATTESTATION_SCOPE",
        "The managed workload registration does not match this deployment.",
      );
    }
    this.#registration.workloadState.persistRegistrationReceipt(receipt);
    this.#registration.registered = true;
    this.#registration.projectTokenProvider = null;
  }

  #recordFailure(error) {
    if (error.code === "INCOMPLETE_CANDIDATE") {
      this.#increment("candidate_incomplete");
    } else if (error.retryable) {
      this.#increment("candidate_delivery_expired");
    } else {
      this.#increment("candidate_rejected");
    }
  }

  #increment(key) {
    this.#recall[key] = Math.min(COUNTER_MAXIMUM, this.#recall[key] + 1);
  }
}

const constructionToken = Symbol("reproit-managed-sink-construction");

function validateConfiguration(configuration) {
  if (
    !isObject(configuration) ||
    typeof configuration.captureSignerId !== "string" ||
    configuration.captureSignerId.length === 0 ||
    configuration.captureSignerId.length > 256 ||
    !(configuration.captureSignerPublicKey instanceof Uint8Array) ||
    configuration.captureSignerPublicKey.length !== 32
  ) {
    throw schemaInvalid();
  }
  requireTypedId(configuration.serviceId, "service_id");
}

function projectTokenProvider(configuration) {
  if (typeof configuration.projectTokenProvider === "function") {
    return configuration.projectTokenProvider;
  }
  if (configuration.projectToken !== null && configuration.projectToken !== undefined) {
    return () => configuration.projectToken;
  }
  return null;
}

function validateUnsignedDeployment(deployment) {
  const repositoryId = deployment.repository_id;
  const runtimeEndpoint = deployment.runtime_endpoint;
  const servicePath = deployment.service_path;
  const sourceRevision = deployment.source_revision;
  if (
    deployment.format !== "reproit.deployment.v1" ||
    deployment.processing_mode !== "managed" ||
    !validTypedId(deployment.organization_id, "organization_id") ||
    !validTypedId(deployment.project_id, "project_id") ||
    !validTypedId(deployment.service_id, "service_id") ||
    typeof repositoryId !== "string" ||
    repositoryId.length < 1 ||
    repositoryId.length > 256 ||
    typeof runtimeEndpoint !== "string" ||
    runtimeEndpoint.length < 1 ||
    runtimeEndpoint.length > 2_048 ||
    typeof servicePath !== "string" ||
    servicePath.startsWith("/") ||
    servicePath.split("/").some((part) => part === "..") ||
    typeof sourceRevision !== "string" ||
    sourceRevision.length < 1 ||
    sourceRevision.length > 256
  ) {
    throw schemaInvalid();
  }
  validateCapabilities(deployment.runtime_capabilities);
}

function prepareManagedDeployment(deployment, manifest, serviceId) {
  if (deployment.service_id !== serviceId) {
    throw new ManagedError(
      "AUTHORIZATION_DENIED",
      "The managed deployment belongs to a different service.",
    );
  }
  deployment.processing_mode = "managed";
  deployment.subject = subjectBinding(manifest);
  const capabilities = [...(deployment.runtime_capabilities ?? [])];
  capabilities.push(manifest.architecture, manifest.operating_system);
  // The candidate includes the process-visible processor requirements.
  capabilities.push(...captureProcessorCapabilities());
  deployment.runtime_capabilities = [...new Set(capabilities)].sort();
  deployment.signer_key_id = "";
  deployment.signature = "";
}

export function managedDeploymentBindingDigest(deployment) {
  const binding = {
    format: deployment.format,
    organization_id: deployment.organization_id,
    processing_mode: deployment.processing_mode,
    project_id: deployment.project_id,
    repository_id: deployment.repository_id,
    runtime_capabilities: deployment.runtime_capabilities,
    runtime_endpoint: deployment.runtime_endpoint,
    service_id: deployment.service_id,
    service_path: deployment.service_path,
    source_revision: deployment.source_revision,
    subject: deployment.subject,
  };
  return canonicalDigest(binding);
}

function registrationTokenRequired() {
  return new ManagedError(
    "AUTHENTICATION_REQUIRED",
    "The managed project token is required for first registration.",
  );
}

// Mirror the reproit-core Deployment::validate checks the SDK can prove.
function validateDeployment(deployment) {
  validateUnsignedDeployment(deployment);
  const repositoryId = deployment.repository_id;
  const runtimeEndpoint = deployment.runtime_endpoint;
  const servicePath = deployment.service_path;
  const signerKeyId = deployment.signer_key_id;
  const sourceRevision = deployment.source_revision;
  if (
    deployment.format !== "reproit.deployment.v1" ||
    typeof repositoryId !== "string" ||
    repositoryId.length < 1 ||
    repositoryId.length > 256 ||
    typeof runtimeEndpoint !== "string" ||
    runtimeEndpoint.length < 1 ||
    runtimeEndpoint.length > 2_048 ||
    typeof servicePath !== "string" ||
    servicePath.length === 0 ||
    servicePath.startsWith("/") ||
    servicePath.split("/").some((part) => part === "..") ||
    typeof signerKeyId !== "string" ||
    signerKeyId.length < 1 ||
    signerKeyId.length > 256 ||
    typeof sourceRevision !== "string" ||
    sourceRevision.length < 1 ||
    sourceRevision.length > 256 ||
    !validTimestamp(deployment.signed_at)
  ) {
    throw schemaInvalid();
  }
  validateCapabilities(deployment.runtime_capabilities);
  decodeBase64url(deployment.signature, 64);
}

File: sdks/python/reproit_sdk/managed_sink.py
"""Bounded managed candidate sink with fail-open delivery.

Mirrors crates/reproit-sdk-rust/src/managed_sink.rs: a bounded in-process
queue, one background delivery worker, recall counters without customer
values, and fail-open semantics. A managed SDK failure never changes the
application's behavior.
"""

from __future__ import annotations

import copy
import json
import os
import queue
import threading
import time
from collections.abc import Callable, Mapping
from dataclasses import dataclass

from reproit_sdk import (
    MAX_GLOBAL_BYTES,
    MAX_OPERATION_BYTES,
    MAX_QUEUED_CANDIDATES,
    canonical_bytes,
    _PROCESS_RESOURCES,
)

from .managed_candidate import (
    FrozenManagedCaptureClosure,
    ManagedCaptureClosure,
    PreparedManagedCandidate,
)
from .managed_identity import (
    ManagedWorkloadIdentityState,
    ManagedWorkloadRegistrationReceipt,
    managed_deployment_binding_digest,
    managed_workload_key_id,
)
from .managed_protocol import (
    ManagedError,
    canonical_digest,
    decode_base64url,
    encode_base64url,
    now_timestamp,
    require_typed_id,
    schema_invalid,
    sign_bytes,
    valid_timestamp,
    validate_capabilities,
    verification_key,
    verify_signed_value,
)
from .managed_subject import (
    PythonSubjectPackage,
    package_running_python_subject,
    subject_binding,
)
from .managed_transport import ManagedProjectToken
from .processor_capture import capture_processor_capabilities

REGISTRATION_TIMEOUT_SECONDS = 5.0
CANDIDATE_DELIVERY_LIFETIME_SECONDS = 1_800.0
_COUNTER_MAXIMUM = (1 << 63) - 1
_RECALL_KEYS = (
    "candidate_delivery_expired",
    "candidate_durably_accepted",
    "candidate_incomplete",
    "candidate_queue_full",
    "candidate_rejected",
    "eligible_failure_observed",
    "suppressed_exact_storm",
    "suppressed_high_cardinality_storm",
)


@dataclass(frozen=True)
class ManagedSinkConfiguration:
    capture_signer_id: str
    capture_signer_public_key: bytes
    project_token: ManagedProjectToken | Callable[[], ManagedProjectToken] | None
    service_id: str
    workload_state_root: str


class _DeadlineManagedClient:
    def __init__(self, client: object, deadline: float):
        self._client = client
        self._deadline = deadline

    def _timeout(self, requested: float) -> float:
        remaining = self._deadline - time.monotonic()
        if remaining <= 0:
            raise ManagedError(
                "SERVICE_UNAVAILABLE",
                "The candidate delivery lifetime expired.",
            )
        return min(requested, remaining)

    def _check(self) -> None:
        self._timeout(0.001)

    def register_workload_key(self, token, request, timeout):
        result = self._client.register_workload_key(
            token, request, self._timeout(timeout)
        )
        self._check()
        return result

    def request_encryption_grant(self, request, timeout):
        result = self._client.request_encryption_grant(
            request, self._timeout(timeout)
        )
        self._check()
        return result

    def start(self, request, timeout):
        result = self._client.start(request, self._timeout(timeout))
        self._check()
        return result

    def missing(self, upload_id, upload_token, cursor, timeout):
        result = self._client.missing(
            upload_id, upload_token, cursor, self._timeout(timeout)
        )
        self._check()
        return result

    def upload_object(self, upload_url, digest, value, timeout):
        self._client.upload_object(
            upload_url, digest, value, self._timeout(timeout)
        )
        self._check()

    def commit(self, upload_id, upload_token, timeout):
        result = self._client.commit(
            upload_id, upload_token, self._timeout(timeout)
        )
        self._check()
        return result

    def cancel(self, upload_id, upload_token, timeout):
        result = self._client.cancel(
            upload_id, upload_token, self._timeout(timeout)
        )
        self._check()
        return result


class ManagedCandidateSink:
    """Deliver complete managed candidates through the bounded upload session."""

    def __init__(
        self,
        client: object,
        closure: ManagedCaptureClosure | FrozenManagedCaptureClosure,
        configuration: ManagedSinkConfiguration,
        subject: PythonSubjectPackage | None = None,
        operation_id: str | None = None,
    ):
        _validate_configuration(configuration)
        if isinstance(closure, ManagedCaptureClosure):
            closure = FrozenManagedCaptureClosure(closure)
        if operation_id is not None:
            require_typed_id(operation_id, "operation_id")
        if subject is None:
            subject = package_running_python_subject()
        self._client = client
        self._closure = closure
        self._configuration = configuration
        self._operation_id = operation_id
        self._subject = subject
        self._world_id = closure.world_id()
        self._project_token = configuration.project_token
        self._workload_identity: ManagedWorkloadIdentityState | None = None
        self._workload_signing_key: bytes | None = None
        self._workload_public_key: bytes | None = None
        self._workload_key_id: str | None = None
        self._registration_request: dict[str, object] | None = None
        self._registration_receipt: ManagedWorkloadRegistrationReceipt | None = None
        self._registration_lock = threading.Lock()
        self._lock = threading.Lock()
        self._idle = threading.Condition(self._lock)
        self._queued_bytes = 0
        self._queued_candidates = 0
        self._active = False
        self._recall = {key: 0 for key in _RECALL_KEYS}
        self._queue: queue.Queue[tuple[dict[str, object], int, float]] = queue.Queue(
            MAX_QUEUED_CANDIDATES
        )
        ready = threading.Event()
        worker = threading.Thread(
            target=self._worker,
            args=(ready,),
            name="reproit-managed-python-capture",
            daemon=True,
        )
        worker.start()
        if not ready.wait(REGISTRATION_TIMEOUT_SECONDS):
            raise ManagedError(
                "SERVICE_UNAVAILABLE", "The managed capture worker could not start."
            )

    @property
    def processing_modes(self) -> frozenset[str]:
        return frozenset(("managed",))

    @property
    def queued_bytes(self) -> int:
        with _PROCESS_RESOURCES.lock:
            return _PROCESS_RESOURCES.queued_bytes

    @property
    def recall_counters(self) -> dict[str, int]:
        """Return bounded counters that contain no customer values."""
        with self._lock:
            return dict(self._recall)

    @property
    def subject_manifest(self) -> dict[str, object]:
        return self._subject.manifest

    @property
    def workload_key_id(self) -> str:
        if self._workload_key_id is None:
            raise ManagedError(
                "CONFIG_CONFLICT", "The managed deployment is not bound."
            )
        return self._workload_key_id

    @property
    def workload_public_key(self) -> bytes:
        if self._workload_public_key is None:
            raise ManagedError(
                "CONFIG_CONFLICT", "The managed deployment is not bound."
            )
        return self._workload_public_key

    @property
    def world_id(self) -> str:
        return self._world_id

    def bind_deployment(self, deployment: dict[str, object]) -> None:
        """Bind the deployment to this subject and sign it as this workload."""
        if deployment.get("service_id") != self._configuration.service_id:
            raise ManagedError(
                "AUTHORIZATION_DENIED",
                "The managed deployment belongs to a different service.",
            )
        manifest = self._subject.manifest
        deployment["processing_mode"] = "managed"
        deployment["subject"] = subject_binding(manifest)
        capabilities = list(deployment.get("runtime_capabilities", []))
        capabilities.extend((manifest["architecture"], manifest["operating_system"]))
        # The captured World's process-visible processor view travels with
        # the candidate so admission starts from the complete observation
        # (spec 7.8.1).
        capabilities.extend(capture_processor_capabilities())
        deployment["runtime_capabilities"] = sorted(set(capabilities))
        deployment["signer_key_id"] = ""
        deployment["signature"] = ""
        binding_digest = managed_deployment_binding_digest(deployment)
        identity = ManagedWorkloadIdentityState.from_state_root(
            self._configuration.workload_state_root, binding_digest
        )
        signing_key = identity.load_or_create_key()
        deployment["signed_at"] = identity.load_or_create_deployment_signed_at(
            binding_digest, deployment.get("signed_at")
        )
        public_key = verification_key(signing_key)
        key_id = managed_workload_key_id(public_key)
        deployment["signer_key_id"] = key_id
        deployment["signature"] = sign_bytes(canonical_bytes(deployment), signing_key)
        _validate_deployment(deployment)
        request = {
            "algorithm": "Ed25519",
            "deployment": copy.deepcopy(deployment),
            "public_key": encode_base64url(public_key),
            "service_id": self._configuration.service_id,
        }
        receipt = ManagedWorkloadRegistrationReceipt(
            deployment_digest=canonical_digest(deployment),
            service_id=self._configuration.service_id,
            workload_key_id=key_id,
        )
        if (
            self._registration_request is not None
            and self._registration_request != request
        ):
            raise ManagedError(
                "CONFIG_CONFLICT",
                "The managed sink is already bound to another deployment.",
            )
        self._workload_identity = identity
        self._workload_signing_key = signing_key
        self._workload_public_key = public_key
        self._workload_key_id = key_id
        self._registration_request = request
        self._registration_receipt = receipt

    def wait_until_idle(self, timeout: float) -> bool:
        deadline = time.monotonic() + timeout
        with self._idle:
            while self._active or self._queued_candidates != 0:
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    return False
                self._idle.wait(remaining)
            return True

    def try_send(self, capture_id: str, candidate: bytes) -> bool:
        """Queue one complete candidate. Never raise into the application."""
        try:
            value = self._authorized_candidate(capture_id, candidate)
        except Exception:
            self._increment("candidate_incomplete")
            return False
        if not _PROCESS_RESOURCES.reserve_candidate(len(candidate)):
            with self._lock:
                self._recall["candidate_queue_full"] = min(
                    _COUNTER_MAXIMUM, self._recall["candidate_queue_full"] + 1
                )
            return False
        with self._lock:
            self._queued_bytes += len(candidate)
            self._queued_candidates += 1
        try:
            self._queue.put_nowait((value, len(candidate), time.monotonic()))
            return True
        except queue.Full:
            with self._lock:
                self._queued_bytes -= len(candidate)
                self._queued_candidates -= 1
                self._recall["candidate_queue_full"] = min(
                    _COUNTER_MAXIMUM, self._recall["candidate_queue_full"] + 1
                )
            _PROCESS_RESOURCES.release_candidate(len(candidate))
            return False

    def _authorized_candidate(
        self, capture_id: str, candidate: bytes
    ) -> dict[str, object]:
        if len(candidate) > MAX_OPERATION_BYTES:
            raise schema_invalid()
        value = json.loads(candidate)
        if (
            not isinstance(value, Mapping)
            or canonical_bytes(value) != candidate
            or value.get("capture_id") != capture_id
            or value.get("processing_mode") != "managed"
        ):
            raise schema_invalid()
        deployment = value.get("deployment")
        if (
            not isinstance(deployment, Mapping)
            or deployment.get("processing_mode") != "managed"
            or deployment.get("service_id") != self._configuration.service_id
            or self._workload_key_id is None
            or self._workload_public_key is None
            or deployment.get("signer_key_id") != self._workload_key_id
            or (
                self._operation_id is not None
                and value.get("operation_id") != self._operation_id
            )
        ):
            raise ManagedError(
                "AUTHORIZATION_DENIED",
                "The managed deployment does not use the registered workload key.",
            )
        verify_signed_value(deployment, self._workload_public_key)
        return dict(value)

    def _worker(self, ready: threading.Event) -> None:
        ready.set()
        while True:
            value, size, queued_at = self._queue.get()
            try:
                if time.monotonic() - queued_at >= CANDIDATE_DELIVERY_LIFETIME_SECONDS:
                    self._increment("candidate_delivery_expired")
                    continue
                with self._lock:
                    self._active = True
                try:
                    self._deliver(
                        value,
                        queued_at + CANDIDATE_DELIVERY_LIFETIME_SECONDS,
                    )
                except ManagedError as error:
                    self._record_failure(error)
                except Exception:
                    self._increment("candidate_rejected")
                else:
                    self._increment("candidate_durably_accepted")
            finally:
                with self._idle:
                    self._active = False
                    self._queued_bytes = max(0, self._queued_bytes - size)
                    self._queued_candidates = max(0, self._queued_candidates - 1)
                    self._idle.notify_all()
                _PROCESS_RESOURCES.release_candidate(size)
                self._queue.task_done()

    def _deliver(self, candidate: Mapping[str, object], deadline: float) -> None:
        configuration = self._configuration
        client = _DeadlineManagedClient(self._client, deadline)
        prepared = PreparedManagedCandidate.prepare_complete(
            candidate, self._subject, self._closure
        )
        self._ensure_registered(client)
        if self._workload_key_id is None or self._workload_signing_key is None:
            raise ManagedError(
                "CONFIG_CONFLICT", "The managed deployment is not bound."
            )
        grant = prepared.request_encryption_grant(
            client, self._workload_key_id, self._workload_signing_key
        )
        sealed = prepared.seal(
            grant,
            now_timestamp(),
            configuration.capture_signer_id,
            configuration.capture_signer_public_key,
        )
        renewal = sealed.request_capture_grant_renewal(
            client, self._workload_key_id, self._workload_signing_key
        )
        sealed.apply_renewed_capture_grant(
            renewal,
            now_timestamp(),
            configuration.capture_signer_id,
            configuration.capture_signer_public_key,
        )
        sealed.upload(client)

    def _ensure_registered(self, client: object) -> None:
        with self._registration_lock:
            identity = self._workload_identity
            request = self._registration_request
            receipt = self._registration_receipt
            if identity is None or request is None or receipt is None:
                raise ManagedError(
                    "CONFIG_CONFLICT", "The managed deployment is not bound."
                )
            if identity.load_registration_receipt(receipt) is not None:
                return
            token_source = self._project_token
            if token_source is None:
                raise ManagedError(
                    "AUTHENTICATION_REQUIRED",
                    "A project token is required to register this managed workload.",
                )
            project_token = token_source() if callable(token_source) else token_source
            if not isinstance(project_token, ManagedProjectToken):
                raise ManagedError(
                    "AUTHENTICATION_REQUIRED",
                    "The project token provider did not return a valid project token.",
                )
            registration = client.register_workload_key(
                project_token, request, REGISTRATION_TIMEOUT_SECONDS
            )
            if (
                registration["deployment_digest"] != receipt.deployment_digest
                or registration["key_id"] != receipt.workload_key_id
                or registration["service_id"] != receipt.service_id
            ):
                raise ManagedError(
                    "ATTESTATION_SCOPE",
                    "The managed workload registration does not match this deployment.",
                )
            identity.persist_registration_receipt(receipt)
            self._project_token = None

    def _record_failure(self, error: ManagedError) -> None:
        if error.code == "INCOMPLETE_CANDIDATE":
            self._increment("candidate_incomplete")
        elif error.retryable:
            self._increment("candidate_delivery_expired")
        else:
            self._increment("candidate_rejected")

    def _increment(self, key: str) -> None:
        with self._lock:
            self._recall[key] = min(_COUNTER_MAXIMUM, self._recall[key] + 1)


def _validate_configuration(configuration: ManagedSinkConfiguration) -> None:
    if (
        not isinstance(configuration.capture_signer_id, str)
        or not configuration.capture_signer_id
        or len(configuration.capture_signer_id) > 256
        or not isinstance(configuration.capture_signer_public_key, bytes)
        or len(configuration.capture_signer_public_key) != 32
        or (
            configuration.project_token is not None
            and not isinstance(configuration.project_token, ManagedProjectToken)
            and not callable(configuration.project_token)
        )
        or not isinstance(configuration.workload_state_root, str)
        or not configuration.workload_state_root
        or not os.path.isabs(configuration.workload_state_root)
    ):
        raise schema_invalid()
    require_typed_id(configuration.service_id, "service_id")


def _validate_deployment(deployment: Mapping[str, object]) -> None:
    """Mirror the reproit-core Deployment::validate checks the SDK can prove."""
    repository_id = deployment.get("repository_id")
    runtime_endpoint = deployment.get("runtime_endpoint")
    service_path = deployment.get("service_path")
    signer_key_id = deployment.get("signer_key_id")
    source_revision = deployment.get("source_revision")
    if (
        deployment.get("format") != "reproit.deployment.v1"
        or not isinstance(repository_id, str)
        or not 1 <= len(repository_id) <= 256
        or not isinstance(runtime_endpoint, str)
        or not 1 <= len(runtime_endpoint) <= 2_048
        or not isinstance(service_path, str)
        or not service_path
        or service_path.startswith("/")
        or any(part == ".." for part in service_path.split("/"))
        or not isinstance(signer_key_id, str)
        or not 1 <= len(signer_key_id) <= 256
        or not isinstance(source_revision, str)
        or not 1 <= len(source_revision) <= 256
        or not valid_timestamp(deployment.get("signed_at"))
    ):
        raise schema_invalid()
    validate_capabilities(deployment.get("runtime_capabilities"))
    decode_base64url(deployment.get("signature"), 64)
